docs/source/concepts/pipeline_schema/index.rst

```diff
@@ -44,4 +48,178 @@ The design goals of the pipeline schema are to be:
+If the user wants to, they can override the default implementation.
+
+.. image:: images/02_default_implementation.png
+   :alt: Diagram showing a step with a default implementation
+
+This is useful because some steps in the record linkage process
+are quite straightforward and unlikely to be sites of innovation.
+Default implementations for these allow EasyLink users
+to focus on the more interesting steps
+without taking away any flexibility—
+if there is some unexpected need to use new implementations
+for these less-interesting steps, the user can do that.
+
+Slots
+^^^^^
+
+A **slot** is a semantic type of data that a step either receives or produces.
+
+For example, consider "canonicalization,"
+which is a step in record linkage that produces a "canonical" record for each entity (e.g., each person).
+An implementation of this step might contain a specific algorithm
+for choosing which of the first names
+a person reported in different datasets to use in that person's canonical record.
+This step needs to receive two types of input:
+clusters of record IDs that correspond to the same entity,
+and the full datasets (from before linking) to pull possible values from.
+These two types of input are not interchangeable—
+any implementation of the step needs to know which is which.
+Therefore, they go in different labeled slots.
+
+We draw **input slots** (slots for receiving data) as small circles
+on the border of a step.
+We draw **output slots** (slots for producing data) as small squares
+on the border of a step.
+A step may have multiple of either or both.
+
+.. image:: images/03_slots.png
+   :alt: Diagram showing slots on a step
+
+A **data specification** is a set of rules that data can be validated against.
+Every slot is associated with a data specification
+that any data passing through it must follow.
+It isn't enough to *say* that a particular dataset is "clusters of record IDs"—
+it has to actually *look* how we would expect those to look.
+This could include constraints like having a specific tabular schema,
+uniqueness in certain fields, etc.
+For example, "clusters of record IDs to canonicalize" might entail
+having two columns, :code:`record_id` and :code:`cluster_id`,
+and :code:`record_id` must be unique.
+
+The label on the arrow indicates the data specification
+that the data must follow (this label is implicitly applied to any slots it is connected to).
+The actual description of the data specification is not included in the diagram;
+that will be listed in text below it.
+
+Though we may expand this in the future,
+we currently think of data in terms of *files or directories*.
+Directories may be nested.
+Here are some example specifications:
+
+- A single file in a tabular format with columns A, B, and C.
+- A directory containing three files,
+  where each is in a tabular format and has three columns.
+- A directory containing any number of subdirectories.
+  Each subdirectory must contain two files, where each is
```

Comment:
why two?

Reply:
These are just arbitrary examples. Would it make that clearer if I put them in quotes, or something?